// FluxData/lib_old/entity.js
;!function(exports, undefined) {

	function objectConstructor(EventEmitter2, util, modelCtr){
		function Entity(model, data){
			this._data = {};
			this._model = false;
			this._changedfields;
			
			if(!data){
				//if the model isn't an instance of the Model object, the user supplied the data and no model
				if(!(model instanceof modelCtr.Model)){
					
					data = model;
					model = false;
				}
			}
			
			if(model){
				this.setType(model);
			}
			
			if(data){
				this.suspendChanges = true;
				this.set(data);
				this.suspendChanges = true;
			}
			
			EventEmitter2.call(this, arguments);
		}
		
			util.inherits(Entity, EventEmitter2);
		
			Entity.prototype.add = function(name, type, value){
				this._attributes
			}
			
			Entity.prototype.setType = function(model){
				this._model = model;
			}
			
			Entity.prototype.getType = function(){
				return this._model;
			}
		
			Entity.prototype.get = function(propName){
				if(propName){
					var propParts = false;
					if(propName.indexOf('.')>-1){
						propParts = propName.split('.');
						propName = propParts.shift();
					}
					
					var propAttribute = this._model?this._model.get(propName):false;
					
					if(propAttribute){
						if((propAttribute.type=='entity' || propAttribute.type=='object') && propParts){
							var propValue = propAttribute.get(this._data, this);
							if(propAttribute.type=='object'){
								return getDataValueByString(this._data, propParts.join('.'));
							}else{ //entity
								return propValue.get(propParts.join('.'));	
							}
						}else{
							return propAttribute.get(this._data, this);//use the attribute object to retrievethe value from the data	
						}
					}else{
						if(propParts){
							var propValue = getDataValueByString(this._data, propName);
							
							return getDataValueByString(propValue, propParts.join('.'));
							
						}else{
							return getDataValueByString(this._data, propParts?propParts.join('.'):propName);
						}
							
					}
				}else{
					return this._data;
				}
			}
			
			Entity.prototype.set = function(propName, propValue){
				//TODO: add support for setting nested values
				var self = this;
				if(propValue==undefined){
					propValue = propName;
					propName = false;	
				}
				
				var propParts = false;
				
				if(propName){
					if(propName.indexOf('.')>-1){
						propParts = propName.split('.');
						propName = propParts.shift();
					}
					if(this._model){
						var attr = this._model.get(propName);
						if(attr){
							if(attr.type=='entity' && propParts.length>0){
								//retrieve the entity
								var currentVal = this.get(propName);
								
								currentVal.set(propParts.join('.'), propValue);
								
								this.set(propName, currentVal);
							}else{
								attr.set(this._data, propValue);
								if(!self.suspendChanges){
									this.emit('changed', this, propName);	
								}	
							}	
						}else{
							if(setDataValueByString(this._data, propName, propValue)){
								if(!self.suspendChanges){
									this.emit('changed', this, propName);	
								}	
							}	
						}
					}else{
						if(setDataValueByString(this._data, propName, propValue)){
							if(!self.suspendChanges){
								this.emit('changed', this, propName);	
							}
						}	
					}
				}else{
					//TODO: Need to process all supplied properties against listed attributes
					this._data = propValue;
					if(!self.suspendChanges){
						this.emit('changed', this, propName);	
					}
				}
			}
			
			Entity.prototype.del = function(propName){
				if(propName){
					if(removeDataValueByString(this._data, propName)){
						this.emit('changed', this);
					}
				}else{
					this._data = {};
					this.emit('changed', this);
				}
			}
			
			Entity.prototype.add = function(propName, propValue){
				if(propValue==undefined){
					propValue = propName;
					propName = false;
				}
				if(!propName){
					if(addDataValueByString(this._data, propName, propValue)){
						this.emit('changed', this);
					}
				}else{
					if(Array.isArray(this._data)){
						if(this._data){
							this._data = [this._data];
						}else{
							this._data = [];	
						}
							
					}
					
					this._data.push(propValue);
					this.emit('changed', this);
				}
			}
		
			Entity.prototype.toObject = function(){
				return this._data;
			}
		/*
		 * DATA SUPPORT FUNCTIONS
		 */
		
		function getDataValueByString(data, nameString){
			var self = this;
			if(nameString!=''){
				if(nameString.indexOf('.')>-1){
					var nameParts = nameString.split('.');
					var currentAttr = nameParts.shift();
					var currentValue;
					if(data){
						currentValue = data[currentAttr];	
					}
					
					var newValue = getDataValueByString(currentValue, nameParts.join('.'));
					
					return newValue;
				}else{
					if(data){
						return data[nameString];	
					}else{
						return data;
					}
					
				}
			}else{
				return;	
			}
		}
		
		function setDataValueByString(data, nameString, value){
			var self = this; 
			if(!data){
				data = {};
			}
			if(nameString){
				if(nameString.indexOf('.')>-1){
					var nameParts = nameString.split('.');
					var currentName = nameParts.shift();
					data[currentName] = setDataValueByString(data[currentName], nameParts.join('.'), value);
				}else{
					data[nameString] = value;
				}
			}
			
			return data;
		}
		
		function addDataValueByString(data, nameString, value){
			var self = this; 
			if(!data){
				data = {};
			}
			if(nameString){
				if(nameString.indexOf('.')>-1){
					var nameParts = nameString.split('.');
					var currentName = nameParts.shift();
					data[currentName] = setDataValueByString(data[currentName], nameParts.join('.'), value);
				}else{
					if(!Array.isArray(data[nameString])){
						data[nameString] = [];
					}
					data[nameString].push(value);
				}
			}
			
			return data;
		}
		
		function removeDataValueByString(data, nameString, index){
			var self = this;
			
			if(nameString!=''){
				if(nameString.indexOf('.')>-1){
					var nameParts = nameString.split('.');
					var currentAttr = nameParts.shift();
					var currentValue;
					if(data){
						currentValue = data[currentAttr];	
					}
					
					var newValue = removeDataValueByString(currentValue, nameParts.join('.'));
					
					return newValue;
				}else{
					if(data){
						if(Array.isArray(data[nameString])){
							data[nameString].splice(index, 1);
						}else{
							delete data[nameString];	
						}
						
						return true;	
					}else{
						delete data;
						return true;
					}
					
				}
			}else{
				return;	
			}
		}
	
		return Entity;
	}

	if (typeof define === 'function' && define.amd) {
		define(['/lib/FluxData/node_modules/eventemitter2/lib/eventemitter2.js', '/lib/FluxData/lib/browser_util.js', '/lib/FluxData/lib/model.js'], function(EventEmitter2, util, modelCtr) {
			return objectConstructor(EventEmitter2, util, modelCtr);
		});
	} else {
		
		var EventEmitter2 = require('eventemitter2').EventEmitter2;
		var util = require('util');
		var modelCtr = require(__dirname+'/model.js');
		
		exports.Entity = objectConstructor(EventEmitter2, util, modelCtr);
	}

}(typeof process !== 'undefined' && typeof process.title !== 'undefined' && typeof exports !== 'undefined' ? exports : window);
